File: LeetCode/ProblemSet/003.Longest Substring Without Repeating Characters/longest-substring-without-repeating-characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function (s) {
    let map = {};

    let max = 0, tmpLen = 0, ignore = 0;
    for (let i = 0, len = s.length; i < len; i++) {
        if (map[s[i]] === undefined || map[s[i]] < ignore) {
            map[s[i]] = i;
            tmpLen++;
            max = Math.max(max, tmpLen);
        } else {
            tmpLen = i - map[s[i]];
            ignore = map[s[i]];// 这个前面的忽略
            map[s[i]] = i;
        }
    }
    return max;
};

console.log(lengthOfLongestSubstring('abcabcbb'));
console.log(lengthOfLongestSubstring('bbbb'));
console.log(lengthOfLongestSubstring('pwwkew'));
console.log(lengthOfLongestSubstring("tmmzuxt"));

File: LeetCode/Go-Solutions/ProblemSet/0003.Longest_Substring_Without_Repeating_Characters/longest-substring-without-repeating-characters.go
package _003_Longest_Substring_Without_Repeating_Characters

func lengthOfLongestSubstring(s string) int {
	var max, tmpLen, ignore int
	m := make(map[uint8]int)

	for i, length := 0, len(s); i < length; i++ {
		key := s[i]
		val, ok := m[key]
		if !ok || val < ignore {
			tmpLen++
			if tmpLen > max {
				max = tmpLen
			}
		} else {
			tmpLen = i - val
			ignore = val
		}
		m[key] = i
	}
	return max
}
